feat!: split trajectory classes and implement Cartesian/JointState specializations

bpapaspyros opened the pull request:
<!-- Pull Request guidelines:
1. Give the PR a relevant and descriptive title, using one of the supported commit types:
   [ build, ci, chore, docs, feat, fix, perf, refactor, release, revert, style, test ]
2. Link the PR to the parent issue, which should already describe and motivate the PR
3. Explain how the PR addresses the parent issue
4. Add any supporting resources (screenshots, test results, etc)
5. Help the reviewer by suggesting the method and estimated time to review
6. If applicable, make a checklist of tasks that should be completed before merging
7. If applicable, mention related issues (blocked by / blocks)
8. Post creation actions: tag reviewers and link the PR to its parent issue under the Development section
-->

## Description

<!-- Required: link the parent issue -->

- #216 

Also resolves #217.

<!-- Required: explain how the PR addresses the parent issue -->
This PR solves the issue by splitting the existing Trajectory class to a base and 2 derived classes, namely, CartesianTrajectory and JointTrajectory.

A lot of the functionality of the base class is now marked protected and acts more as a utility for derived classes. The two new classes share similarities but require different handling when it comes to member variables and cleaning up.

**This is missing python bindings for the new types. I believe it will be cleaner to add them separately.**

<!-- Optional: add additional resources (links, screenshots, test results, etc)
## Supporting information
-->

## Review guidelines

<!-- Required: estimate how long a review should take -->
Estimated Time of Review: 20 minutes

<!-- Optional: provide more suggestions such as "Review by commit", "Read this documentation first", etc -->

## Checklist before merging:

- [ ] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

<!-- Optional: define further tasks that should be completed before merging
- [x] Task 1
- [ ] Task 2
-->

<!-- Optional: link related issues
## Related issues
### Blocked by:
- #0

### Blocks:
- #0
-->

bpapaspyros commented on .vscode/c_cpp_properties.json:
```diff
@@ -9,8 +9,8 @@
             "compilerPath": "/usr/bin/gcc",
-            "cStandard": "c17",
-            "cppStandard": "gnu++17",
+            "cStandard": "c23",
+            "cppStandard": "gnu++20",
```
I was expecting to use C++20 features, but it's no longer the case. We can revert or upgrade, no strong preference on my end

bpapaspyros replied:
fair, but we should consider whether we want to do this at some point during this breaking change. (there is still some value in staying back, so that's ok on my end too)

domire8 replied:
Yeah we can do it, but I would just make a separate PR for this

bpapaspyros commented:
One side effect of the new structure is that since we are storing a flat queue but expose `StateType` api, the `[]` operator no longer returns a reference. 

That is, we can no longer do:

```cpp
trajectory[n] = CartesianState(...)
```

Instead, I replaced the functionality by adding `set_point` and `set_points`.

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+   */
+  template<typename DurationT>
+  bool
+  set_point(unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time);
```
For JointState and the IO states, we have the index as last argument of `set_position`/`set_value`. Might be nice to stay consistent

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+  set_point(unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time);
+
+  /**
+   * @brief Set the trajectory point at given index
```
Copy paste

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase() : State() {
+  this->set_type(StateType::NONE);
```
I don't think this is necessary, as long as the constructor is not accessible and we do it correctly in the derived classes

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+
+  if (!this->times_.empty()) {
+    auto const previous_time = this->times_.back();
+    this->times_.push_back(previous_time + new_time);
```
I thought we were not doing that and the times that you provide are times from start, equivalent to JTC?

bpapaspyros replied:
well, we certainly store them as times from start, but the existing Trajectory functions were designed to just take a duration for that point, not time from start. So I kept the convention. 

But I would also prefer that we keep the same conventions across the stack (e.g., with JTC). I can make that change if we all agree

domire8 replied:
hmm it's true that with the add and insert functions, maybe this requires more than I thought. Let me think about it

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+inline bool TrajectoryBase<TrajectoryT>::set_points(
+    const std::vector<TrajectoryT>& points, const std::vector<std::chrono::duration<int64_t, DurationT>>& new_times) {
+  if (points.size() != new_times.size()) {
+    return false;
```
In control libraries we usually throw in these cases, not returning a bool (same above)

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+
+template<typename TrajectoryT>
+inline std::pair<TrajectoryT, std::chrono::nanoseconds> TrajectoryBase<TrajectoryT>::operator[](unsigned int idx) {
+  this->set_empty(false);
```
What is this for?

bpapaspyros replied:
that was legacy code, I assume it was for cases where `trajectory[n]` would have to return a copy of the data and not a reference.

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,165 @@
+#include "state_representation/trajectory/TrajectoryBase.hpp"
+
+namespace state_representation {
+class CartesianTrajectory : public TrajectoryBase<Eigen::VectorXd> {
```
```suggestion
class CartesianTrajectory : public TrajectoryBase<Eigen::Vector<double, 25>> {
```

bpapaspyros replied:
here for us to set the size at compile time we would need to use something like

```cpp
using Vector25d = Eigen::Matrix<double, 25, 1>;
```

At which point the function parameters would either need to be of that type, or we would need to have some special handling of `Eigen::VectorXd` parameters (direct assignments will cause exceptions because on is fixed and the other dynamic)

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,165 @@
+   * @param reference_frame reference frame of the trajectory points
+   * @param name the name of the state
+   */
+  explicit CartesianTrajectory(const std::string& reference_frame = "world", const std::string& name = "");
```
Name is usually first in control libraries. Also it might be neat to have a way of constructing an empty trajectory and then whenever the first point is added, the reference frame of that point is taken as reference frame for the whole trajectory. Not sure if that is possible though.

Additionally, I think a constructor that takes the first point is also reasonable.

bpapaspyros replied:
I added a 2 more constructors to that make use of the `set_point` and `set_points`. I think that should do it

domire8 reviewed (commented):
Review of the headers

bpapaspyros commented on source/state_representation/include/state_representation/StateType.hpp:
```diff
@@ -11,30 +13,31 @@ namespace state_representation {
+  PARAMETER = 14,
+  GEOMETRY_SHAPE = 15,
+  GEOMETRY_ELLIPSOID = 16,
+  CARTESIAN_TRAJECTORY = 17,
+  JOINT_TRAJECTORY = 18,
+  DIGITAL_IO_STATE = 19,
+  ANALOG_IO_STATE = 20,
```
yeah sure

bpapaspyros commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,259 @@
+  /**
+   * @brief Get attribute list of trajectory points
+   */
+  const std::deque<TrajectoryT>& get_points() const;
```
hm, here the thinking was that for proto or some operations we might need to have the flat array. Or even just in case we ever want to use the flat array for whatever reason. 

We are still returning a `const &` so I don't think this would hurt as a public function

bpapaspyros commented:
I have refactored quite a bit following our internal discussion. 

Some comments:
- please check my docstrings, I have moved quite a few things around so mistakes are bound to happen
- I think the tests can be improved and probably some of the tests can be common for all three classes
- I would probably now consider even more shifting to C++20 (even if it's a separate PR) 

This would allow us to constrain the `Trajectory` class to only accept template arguments derived from the `TrajectoryPoint` struct, somewhat protecting/dictating the use of these classes. 

Having said that, it's still not a strict requirement, but if we are to update the versions then it's something to consider for future PRs that relate to this one.

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+
+template<typename TrajectoryT>
+inline const TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) const {
+  return this->points_[index];
```
Does this give `std::out_of_range` if invalid index?

bpapaspyros replied:
yes, but this is always checked in derived classes. Having said that, it's better to move it up here

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+template<typename TrajectoryT>
+inline const std::deque<std::chrono::nanoseconds> TrajectoryBase<TrajectoryT>::get_times_from_start() const {
+  std::deque<std::chrono::nanoseconds> times_from_start;
+  for (unsigned int i = 0; i < this->points_.size(); ++i) {
+    std::chrono::nanoseconds time_from_start = std::chrono::nanoseconds(0);
+    for (unsigned int j = 0; j <= i; ++j) {
+      time_from_start += this->points_[j].duration;
+    }
+    times_from_start.push_back(time_from_start);
+  }
```
```suggestion
  std::chrono::nanoseconds time_from_start = std::chrono::nanoseconds(0);
  for (unsigned int i = 0; i < this->points_.size(); ++i) {
    time_from_start += this->points_[i].duration;
    times_from_start.push_back(time_from_start);
  }
```
or am I missing something?

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+   * @return the trajectory point
+   * @throw std::out_of_range if index is out of range
+   */
+  const TrajectoryT operator[](unsigned int idx) const;
```
```suggestion
  const TrajectoryT operator[](unsigned int index) const;
```

domire8 replied:
```suggestion
  const TrajectoryT& operator[](unsigned int idx) const;
```
or not? Why would you want to return a const object that is not a reference?

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,124 @@
+  explicit CartesianTrajectory(const std::string& name = "", const std::string& reference_frame = "world");
+
+  /**
+   * @brief Constructor with name and reference frame provided
```
And initial point

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,124 @@
+      const std::string& reference_frame = "world");
+
+  /**
+   * @brief Constructor with name and reference frame provided
```
And points

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+    : TrajectoryBase<CartesianTrajectoryPoint>(name), reference_frame_(reference_frame) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+  this->reset();
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = point.get_name();
+  this->TrajectoryBase<CartesianTrajectoryPoint>::add_point(trajectory_point);
```
```suggestion
  this->add_point(point, duration);
```
for the reference frame check

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("The Cartesian state provided is empty");
+  }
+  if (this->get_size() > 0) {
```
I kinda like this but I'm also torn because it might lead to confusion. What if I construct a trajectory with the reference frame and then it's overridden here. Should almost default the reference frame to empty string then instead of world

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  }
+  if (this->get_size() > 0) {
```
That should not happen when you insert no? Or can you insert on an empty trajectory?

bpapaspyros replied:
ok, there a few inconsistencies throughout that I would like to clear up before I attempt to fix.

First, yes, here my intention was to allow inserting on an empty trajectory (that would still be possible like `trajectory.insert_point(some_point, std::chrono::nanoseconds(100), 0);`)

The bigger question (also mentioned in on of your other comments), is how would we like to handle failure/success on empty Trajectories. That is:

- Do we allow add/insert of the first point to change `reference_frame`, `robot_name`, `joint_names` that have been set in the constructor?

On one hand that might be convenient, on the other hand it might cause confusion quite fast. Although that's not what I have currently implemented, I think that perhaps failing is better. 

Then, however, some behaviours would have to be utilized by first invoking a `reset`. That would also create some question about design choices around the default value of `reference_frame` in `CartesianTrajectory` that you also raised. Our Cartesian states default to `world` and I thought it would be sensible to keep the convention here (but then `reset` would revert to `world`?). However, that would mean that we require the user to really set the `reference_frame` on construction if they don't want to fail the insert/add because of a potential difference there that was now (almost) taken care off for them.

domire8 replied:
Ok so I think I would like to go like that:

### Cartesian

1. Emtpy constructor: no name, reference frame default to world and will stay there (consistent with the rest of the stack)
2. Constructor with name and reference frame
3. Constructor with name, point, duration (in that order): reference frame taken from point
4. Constructor with name, vector of points, vector of durations (in that order): verify all reference frames are the same, set reference frame from that

- get_point, add_point, add_points, get_points, insert_points do the obvious checks
- reset is here to reset a state to empty and have no data, it doesn't mean we reset the reference frames or joint frames. hence, for me, reset should just do that, not mess with reference frame
- set_points is a bit more tricky, I would say that the size of vectors passed as arguments need to correspond to the existing size of the trajectory. If one wants to change all points and size on a trajectory object, one can just call `reset` and then `add_points`.
- If we really want to go all the way, we could introduce a `set_reference_frame(CartesianPose pose)` that multiplies all points with the pose in order to actually perform a change of reference frame and not just a renaming.

### Joint
all the same, except that we can probably allow a `set_joint_names(std::vector<std::string>)` directly.


What do you think?

bpapaspyros replied:
I agree on all points, with only one potential difference

> * reset is here to reset a state to empty and have no data, it doesn't mean we reset the reference frames or joint frames. hence, for me, reset should just do that, not mess with reference frame

It seems to be that it's more semantically correct for `reset` to bring us back to what, for example, the empty constructor would do. What you describe here was closer to what I wanted the `clear` to do, that is, simply remove all points but maintain a check on `reference_frame` (or the joint-specific variables).

So, in my mind, reset should really set the reference frame back to `world` in this case. Would it make sense for you to bring back the `clear` method, or do you think reset is still ok semantically if we follow your suggestion? I am leaning on having the clear back.

domire8 replied:
Hmm while your intuition is in line with the docstring of [State::reset](https://github.com/aica-technology/control-libraries/blob/f72569efbfd7db623e0665b12a01274b5f9f11e8/source/state_representation/include/state_representation/State.hpp#L132), it would be really inconsistent with all other classes where `reset` only resets timestamp, emptiness, data, and not joint names or reference frame

bpapaspyros replied:
oh, indeed I didn't realize that's not precisely the way we use it. Ok, let me get started with this in mind then

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  }
+  if (point.get_reference_frame() != reference_frame_) {
```
```suggestion
  if (point.get_reference_frame() != this->reference_frame_) {
```

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+std::pair<CartesianState, const std::chrono::nanoseconds> CartesianTrajectory::operator[](unsigned int idx) const {
+  try {
+    auto point = this->TrajectoryBase<CartesianTrajectoryPoint>::operator[](idx);
+    CartesianState state;
+    state.set_data(point.data);
+    state.set_name(point.name);
+    state.set_reference_frame(this->reference_frame_);
```
```suggestion
    CartesianState state(point.name, this->reference_frame_);
    state.set_data(point.data);
```

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,190 @@
+
+void CartesianTrajectory::delete_point() {
+  this->TrajectoryBase<CartesianTrajectoryPoint>::delete_point();
+  if (this->get_size() == 0) {
```
```suggestion
  if (this->get_size() != 0) {
```

bpapaspyros commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+   * @brief Get attribute list of trajectory point durations
+   * @return the list of trajectory point durations
+   */
+  const std::deque<std::chrono::nanoseconds> get_durations() const;
```
I also don't have a strong opinion here. In practice `std::deque` is better when we are dynamically adding elements all the time, but `std::vector` is more efficient for most operations as soon as its size is fixed. 

In that sense, if we assume/anticipate that on average our users would create a trajectory and not keep adding/removing elements all the time, it might be even better to consider storing everything in vectors.

domire8 replied:
How do you feel about storing deque but returning vector?

bpapaspyros commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+class TrajectoryBase : public State {
+public:
+  /**
+   * @brief Get attribute list of trajectory point durations
```
no specific reason here, just my headspace at the time. I agree we can go more direct

bpapaspyros commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,240 @@
+  if (points.size() == 0) {
+    return;
+  }
+  this->clear();
```
hm ok. Here I though that if one provides the points and calls the `set_points` they would assume that the trajectory is entirely composed of those points. So in that sense, yes the trajectory size should/would change.

I did now add an `add_points` function too, which would simply append.

Would that be sensible?

domire8 replied:
:+1: I think so yes

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,295 @@
+    this->points_.pop_back();
+  }
+  if (this->points_.empty()) {
+    this->set_empty(false);
```
```suggestion
    this->set_empty(true);
```

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,295 @@
+  }
+  this->points_.erase(this->points_.begin() + index);
+  if (this->points_.empty()) {
+    this->set_empty(false);
```
```suggestion
    this->set_empty(true);
```

bpapaspyros commented:
@domire8 A few things I want to go over in case you review before that

- I want to think a bit more about my constructors. They currently work and pass the tests I wrote, but we might be able to optimize more
- The exceptions need to be cleaned up a bit and perhaps refactored so that I don't always have to call functions that check for them multiple times
- I now have setters for the `reference_frame`, `joint_names`, and `robot_name`, but the current design makes it so that:

`reset()` -> clears points, doesn't touch other member vars

Let's take the `CartesianTrajectory`, that also means that for you to add points that now have a different reference frame you would have to go:

```cpp
trajectory.set_reference_frame(some_pose)

// if new points different in size
trajectory.reset()
trajectory.add_points(points)

// if new points same in size
trajectory.set_points(points) // here we could set the points and then the reference frame 
```

For Joint trajectories, things are a bit more direct and the setters will directly change the joint names or robot name directly without checking something


- `add_point`
  - `CartesianTrajectory`: will fail if the first point to be added has a different or non-default reference frame from what was set in the constructor even if the queue is empty. Problem is, we can't use `set_reference_frame` if the queue is empty because that internally calls `set_points`. I can probably just bypass this and copy the reference frame but it seems awkward to have to provide a `CartesianPose`
  - `JointTrajectory`: will fail unless explicitly calling `set_joint_names` or `set_robot_name` even if those are empty due to the construction.

domire8 commented:
Regarding the constructors, I'd also like to see for consistency
- Empty constructor `CartesianTrajectory()`
- Constructor with name and optional reference frame `CartesianTrajectory(name, reference_frame = "world")`
- Constructor with name, point, duration `CartesianTrajectory(name, point, duration)` NOTE here no reference frame as we take it from the point
- Constructor with name, points, durations. NOTE here also no reference frame as we compare all points and accept if all the same.

I see that adding new points with different reference frame requires multiple function calls, but in that case the user should just create a new object honestly. I feel that it's quite clear if the reference frame is set on construction, and if you want to change it you need to call `set_reference_frame` with a pose and it will be applied to all points. If there are no points in the queue and one calls `set_reference_frame` we could even throw an exception because it doesn't make sense.

Also, I'm not sure that the `robot_name` makes sense here at all. We will want to add `JointState` objects that don't necessarily have the same name (`config_1`, `config_2` etc.) so you won't be able to check for compatibility. I think we can only use joint names for compatiblity.

bpapaspyros commented:
> Regarding the constructors, I'd also like to see for consistency
> 
> * Empty constructor `CartesianTrajectory()`
> * Constructor with name and optional reference frame `CartesianTrajectory(name, reference_frame = "world")`
> * Constructor with name, point, duration `CartesianTrajectory(name, point, duration)` NOTE here no reference frame as we take it from the point
> * Constructor with name, points, durations. NOTE here also no reference frame as we compare all points and accept if all the same.
> 
> I see that adding new points with different reference frame requires multiple function calls, but in that case the user should just create a new object honestly. I feel that it's quite clear if the reference frame is set on construction, and if you want to change it you need to call `set_reference_frame` with a pose and it will be applied to all points. If there are no points in the queue and one calls `set_reference_frame` we could even throw an exception because it doesn't make sense.
> 
> Also, I'm not sure that the `robot_name` makes sense here at all. We will want to add `JointState` objects that don't necessarily have the same name (`config_1`, `config_2` etc.) so you won't be able to check for compatibility. I think we can only use joint names for compatiblity.

Noted, thanks for checking. 

Regarding the robot name, my understanding was this should be common for all `JointState`s (although it wouldn't explicitly cause an issue). If we allow different ones, that also means we should be keeping them in their own vector to be able to reconstruct the `JointState`s. Sounds ok to you?

domire8 commented:
> If we allow different ones, that also means we should be keeping them in their own vector to be able to reconstruct the JointStates. Sounds ok to you?

Yes I think so, same as for the Cartesian

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,301 @@
+  /**
+   * @brief Empty constructor
+   */
+  explicit TrajectoryBase();
```
```suggestion
  explicit TrajectoryBase() = default;
```
Or do you think the `reset` in line 158 is really necessary?

bpapaspyros replied:
nope, not necessary. Removed

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,301 @@
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase(const std::string& name) : State(name) {
+  this->reset();
```
Not sure if necessary to call reset

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,301 @@
+  if (new_points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  }
+  for (auto point : new_points) {
```
```suggestion
  this->set_empty(false);
  for (auto point : new_points) {
```
It could be empty before

domire8 commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,301 @@
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::insert_point(const TrajectoryT& new_point, unsigned int index) {
+  if (index > this->points_.size()) {
+    throw std::out_of_range("Index out of range");
+  }
```
```suggestion
  this->assert_index_in_range(index);
```
What we have done in other places is to wrap that check that we have several times in a helper. https://github.com/aica-technology/control-libraries/blob/f72569efbfd7db623e0665b12a01274b5f9f11e8/source/state_representation/include/state_representation/IOState.hpp#L142

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+ * @class CartesianTrajectoryPoint
+ * @brief Struct to represent a Cartesian trajectory point
+ */
+struct CartesianTrajectoryPoint : public TrajectoryPoint {};
```
Is this necessary or just nice? I'm just wondering. And while we are here, you might as well define a constructor from a duration and CartesianState, that will help reduce code in the source file

bpapaspyros replied:
Yes, I think this is just nicer to have a dedicates struct for this class which could also allow us to more easily specify fields if ever needed.

Also, my original understanding (because of the attribute name) was that JointStates have a robot name, not necessarily a state name, that is common to all. In that case, we would only need to store a name for Cartesian states. But still, I find it a bit nicer to keep it there for future use

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+  explicit CartesianTrajectory(const std::string& name, const std::string& reference_frame = "world");
+
+  /**
+   * @brief Constructor with initial point, duration, name, and reference frame provided
```
Outdated docstring

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+   * @param duration the initial duration
+   * @param name the name of the state
+   * @throw EmptyStateException if point is empty
+   * @throw IncompatibleReferenceFramesException if point has different reference frame
```
```suggestion
```

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+  );
+
+  /**
+   * @brief Constructor with intial points, durations, name and reference frame provided
```
Outdated

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+  const std::string& get_reference_frame() const;
+
+  /**
+   * @brief Getter of the reference frame as const reference
```
Docstring, this one will have to be a bit more precise in terms of what the effect is

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,145 @@
+   * @throw EmptyStateException if point is empty
+   * @throw IncompatibleReferenceFramesException if point has different reference frame
+   */
+  void add_point(const CartesianState& new_point, const std::chrono::nanoseconds& duration);
```
nit: I would either go for `point` and `duration` or `new_point` and `new_duration` no? rather just `point`

domire8 commented on source/state_representation/include/state_representation/trajectory/JointTrajectory.hpp:
```diff
@@ -0,0 +1,129 @@
+   * @brief Constructor with name and reference frame provided
+   * @param name the name of the state
+   */
+  explicit JointTrajectory(const std::string& name = "");
```
```suggestion
  explicit JointTrajectory();
```

bpapaspyros replied:
here I don't really see why we wouldn't support this. I understand we don't support the construction of an empty `JointState`, but here I think we should stay symmetric with the `Trajectory` classes. 

So, `TrajectoryBase` and `CartesianTrajectory` can both be constructed with a name, I think so should `JointTrajectory`

domire8 replied:
Yes you are right, I was just looking for an empty constructor, which this is I guess. Do you think there is a difference between this constructor and having two constructors, one actually empty and one with the string argument that doesnt have a default argument?

bpapaspyros replied:
in this case I just saw it as extra lines of code that don't differ. We might be more flexible by separating those 2 in case we want to do something else in the empty constructor, but for the time being I don't see the reason to separate them really

domire8 commented on source/state_representation/include/state_representation/trajectory/JointTrajectory.hpp:
```diff
@@ -0,0 +1,129 @@
+  explicit JointTrajectory(const std::string& name = "");
+
+  /**
+   * @brief Constructor with initial point, duration, name, and reference frame provided
```
Docstring

domire8 commented on source/state_representation/include/state_representation/trajectory/JointTrajectory.hpp:
```diff
@@ -0,0 +1,129 @@
+   * @param new_point the new trajectory point
+   * @param duration the duration for the new point
+   * @throw EmptyStateException if point is empty
+   * @throw IncompatibleStatesException if point has different robot name
```
The incompatibility comes from the joint names

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+CartesianTrajectory::CartesianTrajectory(
+    const std::string& name, const CartesianState& point, const std::chrono::nanoseconds& duration
+)
+    : TrajectoryBase<CartesianTrajectoryPoint>(name) {
+  this->set_type(StateType::CARTESIAN_TRAJECTORY);
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("The Cartesian state provided is empty");
+  }
+  this->reference_frame_ = point.get_reference_frame();
+  this->add_point(point, duration);
```
```suggestion
    : CartesianTrajectory(name, point.get_reference_frame()) {
  this->add_point(point, duration);
```

bpapaspyros replied:
Here we actually need to do some checks so that we retrieve the reference frame (or joint names for JointTrajectory)

I am indeed duplicating a bit, but since `add_point` (and related functions) are not allowed to change this variables, we need to duplicate checks in the constructor I think

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+  this->add_point(point, duration);
+}
+
+CartesianTrajectory::CartesianTrajectory(
```
I think this is overly complicated and duplication from `add_points`. I would probably just
```
this->set_type(StateType::CARTESIAN_TRAJECTORY);
if (points.empty()) {
  throw exceptions::IncompatibleSizeException("No points provided");
}
this->reference_frame_ = points[0].get_reference_frame();
this->add_points(points, durations);
```

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+
+void CartesianTrajectory::set_reference_frame(const CartesianPose& pose) {
+  if (pose.is_empty()) {
+    throw exceptions::EmptyStateException("Pose is empty");
```
Not sure this is necessary, it will throw the same exception when the operation happens in line 65

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+  }
+  this->reference_frame_ = pose.get_reference_frame();
+  auto points = this->get_points();
+  for (auto& point : points) {
```
Question: Don't we have a non const getter/operator[] to do that in place? Otherwise I don't understand why we needed to define it in the first place

bpapaspyros replied:
Not for state types, no. We only have a reference on the flat data operator[], the operator[] overloads in derived classes have to create a `std::vector<StateT>` at which point the reference is not something we can use

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (new_points.size() != durations.size()) {
+    throw exceptions::IncompatibleSizeException("The size of the points and durations vectors are not equal");
+  } else if (std::ranges::any_of(new_points, [&](const auto& p) { return p.is_empty(); })) {
```
While the check is correct, I wonder if it's necessary. `add_point` will do the same checks again, and in `add_point` we even provide the names of the incompatible reference frames in the exception message which I find almost more valuable. I'd prefer to get the message from line 79 that tells me where to look instead of the generic message from line 99 that just says that there is at least one problem. In that regard, I would skip the checks of line 96 and 98 but that's just my opinion

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+  trajectory_point.name = new_point.get_name();
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::insert_point(trajectory_point, index);
+  } catch (...) {
```
What is the difference between just inserting the point and then doing the insertion in a try catch block?

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+void CartesianTrajectory::insert_point(
+    const CartesianState& new_point, const std::chrono::nanoseconds& duration, unsigned int index
+) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (new_point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + new_point.get_reference_frame() + " and " + this->reference_frame_
+    );
```
These lines are also duplicated, which speaks for a static helper up top that does the same
```
assert_valid(const CartesianState& point, const std::string& reference_frame) {
  throw if point empty
  throw if reference frame not the same
}
```

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+  }
+}
+
+void CartesianTrajectory::set_points(
```
There is too much duplication between `set_point` and `set_points`. Either `set_points` has a for loop that calls `set_point` or the only thing that `set_point` does is `this->set_points({point}, {duration});`. My preference is the first

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,210 @@
+  return points;
+}
+
+CartesianState CartesianTrajectory::get_point(unsigned int index) const {
```
Seeing this, I wonder if we miss getters in the base class for `get_duration(unsigned int index)` and `get_time_from_start(unsigned int index)` ?

bpapaspyros replied:
yeah, I think this might be a nice addition for consistency

domire8 commented on source/state_representation/src/trajectory/JointTrajectory.cpp:
```diff
@@ -0,0 +1,184 @@
+void JointTrajectory::add_point(const JointState& new_point, const std::chrono::nanoseconds& duration) {
+  if (new_point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (this->joint_names_ != new_point.get_names()) {
```
Does that operator work to compare all strings from a vector?

bpapaspyros replied:
yes, side note is that in c++20 these operators are automatically generated because of the introduction of the spaceship `<=>` (https://en.cppreference.com/w/cpp/container/vector). TL;DR we are both backwards and forwards compatible and this works

domire8 commented on source/state_representation/src/trajectory/JointTrajectory.cpp:
```diff
@@ -0,0 +1,184 @@
+#include <algorithm>
```
Comments from CartesianTrajectory apply here as well

domire8 reviewed (commented):
Still a few places where I see room for improvement and reduction of code duplication. Happy to talk about it if you disagree with my points

domire8 commented on source/state_representation/include/state_representation/trajectory/CartesianTrajectory.hpp:
```diff
@@ -0,0 +1,174 @@
+   * @param data the (flattened) trajectory data
+   * @param duration the intended duration for the trajectory point 
+   */
+  CartesianTrajectoryPoint(
+      const std::string& name, const Eigen::VectorXd& data, const std::chrono::nanoseconds& duration
+  )
+      : TrajectoryPoint(name, data, duration) {}
```
```suggestion
  CartesianTrajectoryPoint(
      const CartesianState& state, const std::chrono::nanoseconds& duration
  )
      : TrajectoryPoint(state.get_name(), state.data(), duration) {}
```

domire8 commented on source/state_representation/include/state_representation/trajectory/JointTrajectory.hpp:
```diff
@@ -0,0 +1,159 @@
+   * @param data the (flattened) trajectory data
+   * @param duration the intended duration for the trajectory point 
+   */
+  JointTrajectoryPoint(const std::string& name, const Eigen::VectorXd& data, const std::chrono::nanoseconds& duration)
+      : TrajectoryPoint(name, data, duration) {}
```
```suggestion
  JointTrajectoryPoint(const JointState state, const std::chrono::nanoseconds& duration)
      : TrajectoryPoint(state.get_name(), state.data(), duration) {}
```

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,131 @@
+#include "state_representation/exceptions/IncompatibleReferenceFramesException.hpp"
+
+namespace state_representation {
+CartesianTrajectory::CartesianTrajectory() {
```
You don't need to call the parent constructor?

bpapaspyros replied:
in this case I just saw it as extra lines of code that don't differ. We might be more flexible by separating those 2 in case we want to do something else in the empty constructor, but for the time being I don't see the reason to separate them really

domire8 commented on source/state_representation/src/trajectory/CartesianTrajectory.cpp:
```diff
@@ -0,0 +1,131 @@
+  auto point = this->TrajectoryBase<CartesianTrajectoryPoint>::operator[](idx);
+  CartesianState state(point.name, this->reference_frame_);
+  state.set_data(point.data);
+  auto duration = point.duration;
+  return std::make_pair(state, duration);
```
```suggestion
  return std::make_pair(state, point.duration);
```

domire8 reviewed (commented):
:+1:

bpapaspyros commented on source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp:
```diff
@@ -0,0 +1,457 @@
+   * @throw IncompatibleSizeException if points empty
+   */
+  template<typename T>
+  void assert_points_empty(const std::vector<T>& points) const;
```
true

domire8 commented on source/state_representation/include/state_representation/trajectory/JointTrajectory.hpp:
```diff
@@ -0,0 +1,158 @@
+   * @param index the index
+   * @throw std::out_of_range if index is out of range
+   * @throw EmptyStateException if point is empty
+   * @throw IncompatibleStatesException if point has different robot name
```
There are a few places where it's about the "robot name" but that is outdated

domire8 reviewed (commented):
I think we are there now with the classes, they have become very compact with all the assertions and reused code. Nice work

domire8 reviewed (approved):
Very nice tests as well, elegant. Let's get it in and continue from there! :fire:

bpapaspyros commented on source/state_representation/test/tests/test_trajectory.cpp:
```diff
@@ -1,129 +1,377 @@
+  if constexpr (std::is_same_v<PointType, CartesianState>) {
+    EXPECT_STREQ(this->trajectory->get_reference_frame().c_str(), "world");
+  } else if constexpr (std::is_same_v<PointType, JointState>) {
+    EXPECT_NE(this->trajectory->get_joint_names().size(), 0);
```
@domire8 just doing a final pass check and saw this in my tests which means that we don't throw if the trajectory is empty but we ask for the joint names. While this is a bit inconsistent with other getters, in practice we do keep the the list elements for future additions.

What do you think? Is that ok as is for you? I think it is reasonable for this one.

domire8 replied:
It's consistent with the `JointState` class so yeah its ok
